csrc/scheduler/reduction_tma.cpp

```diff
@@ -7,25 +7,139 @@
 // clang-format on
 
 #include <scheduler/reduction_tma.h>
+#include <scheduler/reduction_utils.h>
+#include <scheduler/runtime_info.h>
+#include <scheduler/tools/inlining.h>
+#include <scheduler/utils.h>
 
 namespace nvfuser {
 namespace reduction {
 namespace tma {
 
-std::unique_ptr<ReductionParams> getReductionHeuristics(
+std::unique_ptr<TmaInnerReductionParams> getReductionHeuristics(
     Fusion* fusion,
     SchedulerRuntimeInfo& runtime_info,
     HeuristicDataCache* data_cache) {
   FusionGuard fg(fusion);
-  auto params = std::make_unique<ReductionParams>();
+  auto params = std::make_unique<TmaInnerReductionParams>();
   params->tag = "Reduction TMA heuristics";
-  NVF_THROW("Schedule reduction using TMA");
+  params->cparams.index_type = runtime_info.getIndexType();
   return params;
 }
 
-void scheduleReduction(Fusion* fusion, const ReductionParams* pparams) {
+void scheduleReduction(Fusion* fusion, const TmaInnerReductionParams* pparams) {
   FusionGuard fg(fusion);
-  NVF_THROW("Schedule reduction using TMA");
+
+  // bool isUnrolled = rparams->isUnrolled();
+  bool isUnrolled = true;
+
+  // Cache inputs if unrolled
+  auto cached_inputs = scheduler_utils::cacheInputs(fusion, isUnrolled);
+
+  // Cache and fork outputs
+  auto cached_outputs =
+      scheduler_utils::cacheAndForkOutputs(fusion, isUnrolled);
+
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion);
+
+  scheduler_utils::prepareForMemoryTypePromotion(fusion);
```

Comment:
`clearMemorySpace()` and `prepareForMemoryTypePromotion()` both come from the non-tma reduction scheduler. Are they needed here?

Reply:
Let's leave them here. We can further wrap these common steps into a util funciton shared by tma and non-tma versions.